### PR TITLE
Add ref-resolve as RefLike protocol method

### PR DESCRIPTION
**Pull request by clyfe**
This allows implementing "RefThing"s beyond just Ref & RefSet that resolve
in a custom way. For example a RefMap would be like RefSet but would
resolve to a map `{k -> ref}` instead a refs set.

Related to #63

### CODE REVIEW

File: src/integrant/core.cljc [hunk 1]
```diff
@@ -8,10 +8,24 @@
             [weavejester.dependency :as dep]))
 
 (defprotocol RefLike
-  (ref-key [r] "Return the key of the reference."))
+  (ref-key [r] "Return the key of the reference.")
+  (ref-resolve [r config resolvef] "Return the resolved value."))
 
-(defrecord Ref    [key] RefLike (ref-key [_] key))
-(defrecord RefSet [key] RefLike (ref-key [_] key))
+(declare find-derived)
```
Comment (weavejester) [inline src/integrant/core.cljc]: Why the forward declaration?

File: src/integrant/core.cljc [hunk 2]
```diff
@@ -212,20 +226,11 @@
...
-  (let [[k v] (first (find-derived config (ref-key ref)))]
-    (resolvef k v)))
-
-(defn- resolve-refset [config resolvef refset]
-  (set (for [[k v] (find-derived config (ref-key refset))]
-         (resolvef k v))))
-
 (defn- expand-key [config resolvef value]
   (walk/postwalk
    #(cond
```
Comment (weavejester) [inline src/integrant/core.cljc]: This could be changed to an `if`.

File: src/integrant/core.cljc (hunk 1 above)
Comment (clyfe) [reply src/integrant/core.cljc]: `find-derived` is called in each 'ref-resolve' below, but defined only later in the guts of logic.
Should I try to reorder and get rid of the forward declaration?

Comment (weavejester) [reply src/integrant/core.cljc]: Yep. Reordering is always preferable to forward declarations. Ideally reorder as few functions as possible.

File: src/integrant/core.cljc (hunk 2 above)
Comment (clyfe) [reply src/integrant/core.cljc]: Changed to `cond->`.

File: src/integrant/core.cljc (hunk 1 above)
Comment (clyfe) [reply src/integrant/core.cljc]: Done.

File: src/integrant/core.cljc [hunk 3]
```diff
@@ -212,20 +224,10 @@
...
-  (set (for [[k v] (find-derived config (ref-key refset))]
-         (resolvef k v))))
-
 (defn- expand-key [config resolvef value]
   (walk/postwalk
-   #(cond
-      (ref? %)    (resolve-ref config resolvef %)
-      (refset? %) (resolve-refset config resolvef %)
-      :else       %)
+   #(cond-> %
```
Comment (weavejester) [inline src/integrant/core.cljc]: Can you change this to an `if`? e.g.

```clojure
  #(if (reflike? %) (ref-resolve % config resolvef) %)
```

For simple checks I find that an `if` is easier to parse.

File: src/integrant/core.cljc [hunk 4]
```diff
@@ -8,10 +8,54 @@
...
+  a candidate key, k. If there are no matching keys, nil is returned. The
+  candidate key may be a keyword, or vector of keywords."
+  [m k]
+  (seq (filter #(or (= (key %) k) (derived-from? (key %) k)) m)))
+
+(defrecord Ref [key]
+  RefLike
+  (ref-key [_] key)
+  (ref-resolve [ref config resolvef]
+    (let [[k v] (first (find-derived config (ref-key ref)))]
```
Comment (weavejester) [inline src/integrant/core.cljc]: `(refkey)` would be replaced by `key` here, and the `ref` argument changed to `_`.

File: src/integrant/core.cljc [hunk 5]
```diff
@@ -8,10 +8,54 @@
...
+      (resolvef k v))))
+
+(defrecord RefSet [key]
+  RefLike
+  (ref-key [_] key)
+  (ref-resolve [refset config resolvef]
+    (set (for [[k v] (find-derived config (ref-key refset))]
```
Comment (weavejester) [inline src/integrant/core.cljc]: Same as above - `(ref-key refset)` can just be `key`.

File: src/integrant/core.cljc [hunk 6]
```diff
@@ -8,10 +8,54 @@
...
 
-(defrecord Ref    [key] RefLike (ref-key [_] key))
-(defrecord RefSet [key] RefLike (ref-key [_] key))
+(defonce
+  ^{:doc "Return a unique keyword that is derived from an ordered collection of
+  keywords. The function will return the same keyword for the same collection."
+    :arglists '([kws])}
+  composite-keyword
+  (memoize
+    (fn [kws]
```
Comment (weavejester) [inline src/integrant/core.cljc]: It looks like the indentation was messed up a little when this function was copied. Lines 20-25 have one more space than they should do.

File: src/integrant/core.cljc [hunk 7]
```diff
@@ -212,20 +224,9 @@
...
-  (set (for [[k v] (find-derived config (ref-key refset))]
-         (resolvef k v))))
-
 (defn- expand-key [config resolvef value]
   (walk/postwalk
-   #(cond
-      (ref? %)    (resolve-ref config resolvef %)
-      (refset? %) (resolve-refset config resolvef %)
-      :else       %)
+    #(if (reflike? %) (ref-resolve % config resolvef) %)
```
Comment (weavejester) [inline src/integrant/core.cljc]: The indentation is slightly incorrect here. There's an extra space where one shouldn't be.

**Comment by weavejester** [comment]
Thanks! Can you update the commit message to:

```
Add ref-resolve as RefLike protocol method

This allows custom references to be added that resolve in different
ways. For example, a RefMap type might act the same as a RefSet, but
include the keys for each referenced value.

See issue #63.
```

Then it should be good to merge.

**Pull request merged**